Add service methods for immutable Compute resources and tests

mziccard opened the pull request:
This PR adds spi and service methods for manipulating the following resources:
- DiskType
- MachineType
- License
- Zone
- Region

In c18730c integration tests are also added.

mziccard commented:
I think the coverage drop is the usual one we see when we add the spi layer that is only tested through ITs (which do not concur to compute the coverage).

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java:
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
+      ImmutableList.Builder<DiskType> builder = ImmutableList.builder();
+      Map<String, DiskTypesScopedList> scopedList = aggregatedList.getItems();
+      if (scopedList != null) {
+        for (String key : scopedList.keySet()) {
```
entrySet?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+      this.selector = selector;
+    }
+
+    public String selector() {
```
I am not sure if we discussed that already.
If we don't mind to make this method `public` than we could make all the XXXField enums implement a "Selector" interface and have one static `String selector(Selector required, Selector... other)` method.

mziccard replied:
Yes we discussed it and opened #456 but it seems like we haven't had time to address that issue.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+  enum DiskTypeField {
+    CREATION_TIMESTAMP("creationTimestamp"),
+    DEFAULT_DISK_SIZE_GB("defaultDiskSizeGb"),
+    DESCRIPTION("description"),
```
did you intentionally left out "deprecated"?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+  enum MachineTypeField {
+    CREATION_TIMESTAMP("creationTimestamp"),
+    DESCRIPTION("description"),
+    GUEST_CPUS("cpus"),
```
"cpus" -> "guestCpus"?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+    MEMORY_MB("memoryMb"),
+    NAME("name"),
+    SCRATCH_DISKS("scratchDisks"),
+    DISK_GB("diskGb"),
```
isn't that part of "scratchDisks"?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+  /**
+   * Class for specifying disk type list options.
+   */
+  class DiskTypeListOption extends Option {
```
filter is coming later?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+     */
+    public static RegionListOption fields(RegionField... fields) {
+      StringBuilder builder = new StringBuilder();
+      builder.append("items(").append(RegionField.selector(fields)).append("),nextPageToken");
```
With the "Selectable" interface we could also make fields -> list string a common method...

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+   *
+   * @throws ComputeException upon failure
+   */
+  DiskType getDiskType(DiskTypeId diskTypeId, DiskTypeOption... options) throws ComputeException;
```
No need for "throws ComputeException". Having it in the javadoc is enough.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +30,548 @@
+      throws ComputeException;
+
+  /**
+   * Lists the disk types in the provided zone available to the current project.
```
is the "to the current project" necessary or inferred?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/ComputeImpl.java:
```diff
@@ -16,15 +16,469 @@
+              return serviceOptions.rpc().listDiskTypes(optionsMap);
+            }
+          }, serviceOptions.retryParams(), EXCEPTION_HANDLER);
+      String cursor = result.x();
```
Any reason for including it inside of the retry handling try~catch?

mziccard replied:
No particular reason but this method has the same structure we used almost everywhere else for listing methods.

aozarov replied:
Ok, Fine as is, but I don't think we need to consider that as a recommended way or a pattern we should follow (nor should we go and change other places).

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/ComputeRpc.java:
```diff
@@ -16,5 +16,155 @@
+   *
+   * @throws ComputeException upon failure
+   */
+  DiskType getDiskType(String zone, String diskType, Map<Option, ?> options)
```
same comment about throws.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java:
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
+          .setFields(FIELDS.getString(options))
+          .execute();
+    } catch (IOException ex) {
+      ComputeException serviceException = translate(ex);
```
if you wanted you can make this re-occuring code a generic method that returns T (always null) or throws ComputeException.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java:
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
+    try {
+      DiskTypeList diskTypesList = compute.diskTypes()
+          .list(this.options.projectId(), zone)
+          .setFilter(FILTER.getString(options))
```
Not used yet, right?

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java:
```diff
@@ -43,4 +64,214 @@ public DefaultComputeRpc(ComputeOptions options) {
+
+  @Override
+  public DiskType getDiskType(String zone, String diskType, Map<Option, ?> options)
+      throws ComputeException {
```
No need for the explicit "throws ComputeException" (I see that in Storage we do that in the spi RPC interface and implementation but I think it is a leftover as originally it was IOException). In any case, for consistency I am fine with leaving it as is (here and in the spi interface), but it should be removed from the API interface (that has a javadoc to cover it).

aozarov commented:
Nice work! Done with first iteration. Any reason we don't have codacy analysis here?

mziccard commented:
@aozarov I addressed the comments in 28ed6e8

- Add support for filters: `filter` method to options + filter classes
- Removed unnecessary fields + added missing ones
- Removed redundant "throws" from method signatures
- Add static method in `DefaultComputeRpc` to handle the NOT_FOUND case

I also had to add dedicated option classes for aggregated lists as `fields` option is not completely supported (see #711).

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/compute/Compute.java:
```diff
@@ -24,4 +35,858 @@
+    private final ComparisonOperator operator;
+    private final Object value;
+
+    enum ComparisonOperator {
```
public

aozarov replied:
Oh, looks like you meant it to be package scoped. Fine in that case.

aozarov replied:
The Beta API also supports filtering on multiple expressions. Is it something that you plan to add later?

mziccard replied:
Yes we decided to stick to v1, I plan to support multiple filters as soon as the Beta API becomes v2. This should be quite simple.

aozarov commented on gcloud-java-compute/src/main/java/com/google/gcloud/spi/DefaultComputeRpc.java:
```diff
@@ -43,4 +64,201 @@ public DefaultComputeRpc(ComputeOptions options) {
+          .setFilter(FILTER.getString(options))
+          .setMaxResults(MAX_RESULTS.getLong(options))
+          .setPageToken(PAGE_TOKEN.getString(options))
+          // todo(mziccard): uncomment or remove once #711 is closed
```
unnecessary comment?

mziccard replied:
I forgot to comment the following line.

aozarov commented:
Looks good. Few minor comments. Feel free to merge after they are resolved.
